[RFC] Add support for the Model Checking Intermediate Language (MCIL)

daniel-larraz opened the pull request:
This PR adds MCIL as a version of the smtlib2 language. The current specification of the language can be found [here](https://github.com/ModelChecker/IL/blob/main/description.md).

daniel-larraz committed:
Add support for the Model Checking Intermediate Language (MCIL)

Co-authored-by: kian-weimer <55513580+kian-weimer@users.noreply.github.com>

Gbury commented:
Thank you for the contribution !

I haven't had time to read everything yet, but I already have some questions/remarks:
- for the parsing part, I'd rather avoid using an assoc list for the variables (and the conds), and instead have the `mk_sys` function explicitly take each list separately with a named argument, e.g. : `mk_sys : input:var list -> output:var list -> local:var list -> ... ->`
- it would be nice to have some examples that do pass typing (there are only error examples currently if I'm not mistaken)
- I'm wondering whether we really need to have a separate notion of systems: instead, would it be reasonable to make it so that we have a type (or an indexed family of types) `'local system`, and have system definitions actually create a constant of type `input_var_1_type -> ... -> input_var_n_type -> output_var_1_type -> ... -> output_var_m_type -> (local_var_1_type, ...) system`. That would minimize the required amount of new code (but maybe that would result in worse error messages, so it may not be a good solution...)

daniel-larraz committed:
Make sys_def and sys_check take named arguments instead of assoc list

daniel-larraz committed:
Add missing 'expected' files

daniel-larraz committed:
Add MCIL examples that do pass typing

daniel-larraz commented:
@Gbury , thank you for your quick feedback.

I addressed your first two points.

I'm not sure about the last point. Unless you feel strongly about that change, I would post-pone it.

daniel-larraz committed:
Add MCIL examples with enumeration declarations

Gbury commented:
> I'm not sure about the last point. Unless you feel strongly about that change, I would postpone it.

Agreed, ^^

Another question: why add a dedicated command for declaring enumeration (`declare-enum-sort`) when it can already be done using `declare-datatypes` ?

daniel-larraz commented:
> Another question: why add a dedicated command for declaring enumeration `(declare-enum-sort)` when it can already be done using `declare-datatypes`?

I don't know the answer. I only know that the dedicated command is [part](https://github.com/ModelChecker/IL/blob/main/description.md#enumeration-declaration) of the current MCIL specification. I can ask around to figure out the rationale.

daniel-larraz commented:
> > Another question: why add a dedicated command for declaring enumeration `(declare-enum-sort)` when it can already be done using `declare-datatypes`?
> 
> I don't know the answer. I only know that the dedicated command is [part](https://github.com/ModelChecker/IL/blob/main/description.md#enumeration-declaration) of the current MCIL specification. I can ask around to figure out the rationale.

I asked around.  The `(declare-enum-sort)` command was added because it allows solvers to support enumeration types without having to support the (much more complex) data types.

Gbury reviewed (changes requested):
Overall, the code is mostly good. I did a general overview and left comments to address the major design choices (e.g. having the code checking systems in a separate file instead of in `thf.ml`).

One question I've been pondering is whether we want to keep the name `System`, as it lacks a bit of context to know exactly what it's referring to (in particular in a context that is otherwise usually used for theorem proving). Maybe something that more explicitly mentions model checking would be better.

daniel-larraz committed:
Move checking of transition systems in a separate file

The new code uses the `get_global_custom` and `set_global_custom` functions to
manage the table of defined systems, and calls the check_no_free_wildcards function.
This commit also removes the system constant that is not necessary anymore.
The code that emits warnings for unused variables is enabled again;
no issues were found.

daniel-larraz committed:
Remove unused var from MCIL test

daniel-larraz committed:
Remove unused enum definitions

daniel-larraz committed:
Remove spurious new lines

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
 
-  (* High-level parsing function *)
+  (* MCIL Only System Definitions and Checks *)
   (* ************************************************************************ *)
```
Addressed.

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
+      Was causing errors and its purpose is unknown. *)
+    (* List.iter (check_used_term_var ~kind:`Function_param env) input;
+    List.iter (check_used_term_var ~kind:`Function_param env) output;
+    List.iter (check_used_term_var ~kind:`Function_param env) local; *)
```
I put the code back again, and now it works fine.

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
+
+    let id, f = id_for_sig ssig_env input output local d in
+
+    (* [Kian] Finalize Def -- What do these do? *)
```
Thank you for your explanations! I'll let you know.

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
+    (* [Kian] Finalize Def -- What do these do? *)
+
+    (* let def = parse_def ssig t in
+    let ssig = close_wildcards_in_sig ssig t in
```
Addressed.

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
+
+    (* [Kian] Finalize Def -- What do these do? *)
+
+    (* let def = parse_def ssig t in
```
Got it. Thanks!

daniel-larraz commented on src/typecheck/thf.ml:
```diff
@@ -2527,9 +2551,268 @@ module Make
+
+    (* let def = parse_def ssig t in
+    let ssig = close_wildcards_in_sig ssig t in
+    let id = id_for_def ~freshen:false ~defs:d ~mode tags ssig t in *)
```
Got it. Thanks!

daniel-larraz commented on src/standard/statement.ml:
```diff
@@ -256,6 +288,47 @@ let print_group print fmt ({ contents; recursive; } : _ group) =
+    Format.fprintf fmt "query %a (%a);@ "
+      Id.print name
+      (Misc.print_list ~print_sep:Format.fprintf ~sep:" " ~print:Term.print) formula_names in
+
```
Addressed.

daniel-larraz commented on src/standard/statement.ml:
```diff
@@ -256,6 +288,47 @@ let print_group print fmt ({ contents; recursive; } : _ group) =
+let print_check_sys fmt ({id; loc = _; input; output; local; reachable; assumption; queries}: sys_check) =
+  let print_formula base_name fmt (name, term) = 
+    Format.fprintf fmt "@,%s %a = %a;" base_name Id.print name Term.print term in
+
```
Addressed.

daniel-larraz commented on src/standard/statement.ml:
```diff
@@ -256,6 +288,47 @@ let print_group print fmt ({ contents; recursive; } : _ group) =
+
+  let print_subs fmt subs = 
+    List.iter (Format.fprintf fmt "@,%a" print_sub) subs in
+
```
Addressed.

daniel-larraz commented on src/standard/statement.ml:
```diff
@@ -256,6 +288,47 @@ let print_group print fmt ({ contents; recursive; } : _ group) =
+  print_attrs local
+  (Misc.print_list ~print_sep:Format.fprintf ~sep:"@," ~print:(print_formula "assumption")) assumption
+  (Misc.print_list ~print_sep:Format.fprintf ~sep:"@," ~print:(print_formula "reachable")) reachable
+  (Misc.print_list ~print_sep:Format.fprintf ~sep:"@," ~print:print_query) queries
```
I tried to run `ocp-indent` on the file, but it didn't change anything. Perhaps I'm doing something wrong. In any case, I would suggest to set up `ocamlformat` in the (dune) project to enforce a consistent formatting. How does it sound?

daniel-larraz commented on src/standard/statement.ml:
```diff
@@ -43,6 +43,35 @@ type def = {
+  id      : Id.t;
+  loc : location;
+  values  : Id.t list;
+}
```
I removed it. Thanks!

daniel-larraz committed:
Move Cannot_find_system exception to MCIL module

daniel-larraz committed:
Refine undefined-system error location for sys_check

daniel-larraz committed:
Refine and add errors related to subsystems

daniel-larraz committed:
Update MCIL tests

daniel-larraz committed:
Add error for unbound identifier in query

daniel-larraz committed:
Add duplicate-definition error in MCIL

daniel-larraz committed:
Fix MCIL system_instantiation grammar rule

Gbury commented on src/typecheck/trans_sys.ml:
```diff
@@ -0,0 +1,246 @@
+    parse_subsystems env d
+
+  let finalize_sys (d : Stmt.sys_def) ((env, _), input, output, local) =
+    Type.check_no_free_wildcards env d.init;
```
These functions should be called right after typing the corresponding ast. In essence, wildcards are added to the global `state` stored inside the `env`, hence why the `check_no_free_wildcards` should be called right after typing, to ensure that all free wildcards were actually created when typing some ast.

Gbury commented on src/typecheck/trans_sys.ml:
```diff
@@ -0,0 +1,246 @@
+    Type.check_no_free_wildcards env d.init;
+    Type.check_no_free_wildcards env d.trans;
+    Type.check_no_free_wildcards env d.inv;
+    let input, output, local = vars input, vars output, vars local in
```
This way of writing code is a bit mislead I think. Instead I think it'd be better to 'inline' the call to `vars` for each one of the `List.iter` below.

Gbury commented on src/typecheck/trans_sys.ml:
```diff
@@ -0,0 +1,246 @@
+      if (n1 != n2) then _bad_arity env (Id id) n1 n2 a ;
+      List.iter2
+        (fun (_, v1, _) (_, v2, a2) ->
+          ensure env a2 (Type.T.of_var v2) (Type.T.Var.ty v1)
```
I think that here it would be better to directly check for equality of the types, and raise an error if the equality does not hold.

Gbury commented on src/typecheck/trans_sys.ml:
```diff
@@ -0,0 +1,246 @@
+    envs, parsed_input, parsed_output, parsed_local
+
+  let parse_condition env cond =
+    Type.parse_prop env cond |> ignore
```
This is where you should add the calls to `check_no_free_wildcards` (see my other comment)

Gbury commented on src/typecheck/trans_sys.ml:
```diff
@@ -0,0 +1,246 @@
+          if not (M.mem s cond_ids) then
+            Type._error env (Ast ast) (Type.Cannot_find (s, ""));
+        )
+        | _ -> assert false
```
this should be a proper error, not an `assert false`

Gbury commented on src/typecheck/trans_sys.mli:
```diff
@@ -0,0 +1,26 @@
+  (** Parse a transition system definition *)
+
+  val parse_check : Type.env -> Dolmen.Std.Statement.sys_check -> [> `Sys_check ]
+  (** Parse a transition system check *)
```
As mentionned in other comments: the current types here (and in `parse_def`) will lose information (e.g. the transitions in `parse_def`, and everything in this case), and we should not lose information

Gbury reviewed (changes requested):
Apologies for the very late review.

daniel-larraz commented:
I wasn't able to keep making progress on this due to other higher-priority work. I'm unlikely to have the bandwidth in the near term, so the best way forward may be for someone else to pick it up.

Also, just a note: the language is now called MoXI.
